src/charm.py

```diff
@@ -35,83 +83,577 @@ def __init__(self, *args: typing.Any):
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ops.ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
+            event: Relation joined event.
+        """
+        # rabbitmq charm library doesn't expose the admin user setting
+        if self.unit.is_leader():
+            event.relation.data[self.app]["admin"] = "true"
 
-    @property
-    def _pebble_layer(self) -> pebble.LayerDict:
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
+    def _on_pebble_custom_notice(self, event: ops.PebbleCustomNoticeEvent) -> None:
+        """Handle pebble custom notice event.
+
+        Args:
+            event: Pebble custom notice event.
+        """
+        if event.notice.key.startswith("canonical.com/opencti/"):
+            self._reconcile(event)
+
+    def _reconcile(self, _: ops.EventBase) -> None:
+        """Run charm reconcile function and catch all exceptions."""
+        try:
+            self._reconcile_platform()
+            self.unit.status = ops.ActiveStatus()
+        except (MissingIntegration, MissingConfig, InvalidIntegration, InvalidConfig) as exc:
+            self.unit.status = ops.BlockedStatus(str(exc))
+        except (ContainerNotReady, IntegrationNotReady, PlatformNotReady) as exc:
+            self.unit.status = ops.WaitingStatus(str(exc))
+
+    def _reconcile_platform(self) -> None:
+        """Run charm reconcile function.
+
+        Raises:
+            PlatformNotReady: failed to start the OpenCTI platform at this moment
+        """
+        self._init_peer_relation()
+        self._check_preconditions()
+        health_check_token = self._get_peer_secret(_PEER_SECRET_HEALTH_ACCESS_KEY_SECRET_FIELD)
+        health_check_url = f"http://localhost:8080/health?health_access_key={health_check_token}"
+        self._install_callback_script(health_check_url)
+        self._install_opensearch_cert()
+        self._container.add_layer(
+            "opencti",
+            layer=self._gen_pebble_service_plan(),
+            combine=True,
+        )
+        self._container.replan()
+        self._container.start("platform")
+
+        if not self._is_platform_healthy(health_check_url):
+            self._container.start("charm-callback")
+            raise PlatformNotReady("waiting for opencti platform to start")
```

Comment:
While `PlatformNotReady` is not an error, I feel the naming of the class need to make it clear it is an Exception class.

Reply:
Yes, this concerns me too, but since it's an error, we can't use the `Error` suffix, and the `Exception` suffix is too long. I can't find a perfect solution.